src/app/tabs/camera.cpp

```diff
@@ -0,0 +1,143 @@
+    status2->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+
+    QPushButton *button = new QPushButton("connect", widget);
+    button->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Preferred);
+    button->setFont(Theme::font_14);
+    button->setFlat(true);
+    button->setIconSize(Theme::icon_36);
+    Theme::get_instance()->add_button_icon("wifi", button);
+    connect(button, &QPushButton::clicked, this, [this, stream_url_input](){this->connect_stream(stream_url_input->text());});
+
+    connect(this, &CameraTab::stream_connecting, this, [button, stream_url_input]() {button->setEnabled(false); stream_url_input->setEnabled(false);} );
+    connect(this, &CameraTab::stream_disconnected, this, [button, stream_url_input]() {button->setEnabled(true); stream_url_input->setEnabled(true);} );
+
+    layout->addStretch();
+    layout->addWidget(label, 1);
+    layout->addStretch();
+    layout->addWidget(stream_url_input, 0, Qt::AlignCenter);
+    layout->addWidget(status2, 0, Qt::AlignCenter);
+    layout->addStretch();
+    layout->addWidget(button, 1, Qt::AlignCenter);
+
+    return widget;
+}
+
+void CameraTab::changed_status(QMediaPlayer::MediaStatus mediaStatus)
+{
+    qInfo() << "Camera status changed to: " << mediaStatus;
+
+    switch(mediaStatus)
+    {
+        case QMediaPlayer::InvalidMedia:
+            status2->setText("unable to connect");
+            emit stream_disconnected();
+            break;
+         case QMediaPlayer::LoadingMedia:
+         case QMediaPlayer::LoadedMedia:
+         case QMediaPlayer::BufferedMedia:
+            status2->setText("connecting...");
+            emit stream_connecting();
+            break;
+         case QMediaPlayer::EndOfMedia:
+         case QMediaPlayer::StalledMedia:
+         case QMediaPlayer::NoMedia:
+        default:
+            status2->setText("disconnected");
+            emit stream_disconnected();
+            break;
+    }
+}
+
+void CameraTab::new_metadata(const QString &key, const QVariant &value)
+{
+    status->setText("connected");
+    emit stream_connected();
+}
+
+void CameraTab::disconnect_stream()
+{
+    player->setMedia(QUrl());
+    player->stop();
+}
+
+void CameraTab::connect_stream(QString stream_url)
+{
+    if ( player != NULL )
+        delete player;
+
+    player = new QMediaPlayer;
+    player->setVideoOutput(video_widget);
+
+    connect(player, &QMediaPlayer::mediaStatusChanged, this, &CameraTab::changed_status);
+    connect(player, qOverload<const QString&,const QVariant&>(&QMediaPlayer::metaDataChanged), this, &CameraTab::new_metadata);
+
+    // e.g. "rtsp://10.0.0.185:8554/unicast"
+    QString stream_pipeline = QString("gst-pipeline: rtspsrc location=%1 ! decodebin ! video/x-raw ! videoconvert ! videoscale ! xvimagesink sync=false force-aspect-ratio=false name=\"qtvideosink\"").arg(stream_url);
+    qInfo() << "Playing stream pipeline: " << stream_pipeline;
+    player->setMedia(QUrl(stream_pipeline));
+    player->play();
+}

```

Comment:
newline at end

Reply:
This is what you're looking for, right?

![image](https://user-images.githubusercontent.com/2094779/83805042-26c83800-a67d-11ea-8b6c-6be539058d85.png)

Reply:
(this is the current state of the file but Github shows it differently for whatever reason)

Reply:
Ah yeah i see that, great!